src/common/hist_util.cc

```diff
@@ -785,41 +860,46 @@ void BuildHistSparseKernel(const common::Span<const size_t> rid_span, const floa
       const size_t icol_end_prefect = row_ptr[rid[i+Prefetch::kPrefetchOffset]+1];
 
       PREFETCH_READ_T0(pgh + 2*rid[i + Prefetch::kPrefetchOffset]);
-      for (size_t j = icol_start_prftch; j < icol_end_prefect; j+=Prefetch::kPrefetchStep) {
+      for (size_t j = icol_start_prftch; j < icol_end_prefect; j+=prefetch_step) {
         PREFETCH_READ_T0(gradient_index + j);
       }
     }
-
+    size_t jd = 0;
     for (size_t j = icol_start; j < icol_end; ++j) {
-      const uint32_t idx_bin = 2 * gradient_index[j];
+      const uint32_t idx_bin = 2 * ((uint32_t)(gradient_index[j]) + disps[jd]);
```

Comment:
Probably I missed smth, but I think ```disps[jd]``` is incorrect here, because in sparse case - you have some features missed, but you do increment ```++jd;``` on each iteration - it's not right in general case.

Reply:
fixed